### PR TITLE
Localization tag filtering

### PR DESCRIPTION
**Pull request by mcoughlin**
This PR adds basic localization tag filtering.

### CODE REVIEW

File: skyportal/models/user_notification.py [hunk 1]
```diff
@@ -583,27 +592,41 @@ def receive_after_flush(session, context):
...
+
+                        if "gcn_tags" in pref["gcn_events"].keys():
+                            if len(pref['gcn_events']["gcn_tags"]) > 0:
+                                intersection = list(
+                                    set(event.tags)
+                                    & set(pref['gcn_events']["gcn_tags"])
+                                )
+                                if len(intersection) == 0:
+                                    send_notification = False
+                                    break
```
Comment (guynir42) [inline skyportal/models/user_notification.py:621]: make sure to change all the breaks (if indeed they need to be continue)

File: skyportal/models/user_notification.py [hunk 2]
```diff
@@ -583,27 +592,41 @@ def receive_after_flush(session, context):
...
-                                    if len(intersection) == 0:
-                                        send_notification = False
-                                        break
+                            send_notification = False
+                            break
+                        notice = event.gcn_notices[-1]
+
+                        if "gcn_tags" in pref["gcn_events"].keys():
+                            if len(pref['gcn_events']["gcn_tags"]) > 0:
+                                intersection = list(
```
Comment (guynir42) [inline skyportal/models/user_notification.py:615]: Not sure you need the `list()` here.

File: skyportal/models/user_notification.py [hunk 3]
```diff
@@ -638,32 +653,80 @@ def receive_after_flush(session, context):
...
-                                            user=user,
-                                            text=f"New Notice for GCN Event *{target.dateobs}*, with Notice Type *{gcn.NoticeType(target.notice_type).name}*",
-                                            notification_type="gcn_events",
-                                            url=f"/gcn_events/{str(target.dateobs).replace(' ','T')}",
+                                if len(intersection) == 0:
+                                    send_notification = False
+                                    break
+
+                        if "localization_properties" in pref["gcn_events"].keys():
+                            if len(pref['gcn_events']["localization_properties"]) > 0:
```
Comment (guynir42) [inline skyportal/models/user_notification.py:678]:
```suggestion
                        if pref['gcn_events'].get('localization_properties'):
```

File: skyportal/models/user_notification.py [hunk 4]
```diff
@@ -638,32 +653,80 @@ def receive_after_flush(session, context):
...
+                                )
+                            else:
+                                session.add(
+                                    UserNotification(
+                                        user=user,
+                                        text=f"New GCN Event *{target.dateobs}*, with Notice Type *{gcn.NoticeType(notice.notice_type).name}*",
+                                        notification_type="gcn_events",
+                                        url=f"/gcn_events/{str(target.dateobs).replace(' ','T')}",
                                     )
+                                )
```
Comment (guynir42) [inline skyportal/models/user_notification.py:729]: Lets make the first few words of the `text` input inside the conditional and the rest of the code after it.

**Review by guynir42 — COMMENTED** [review]
I have a lot of comments but mostly they are very minor. 
I would maybe add a test for the localization tags (hopefully just a few lines in an existing test that checks the GCN tags).

File: skyportal/handlers/api/gcn.py [hunk 5]
```diff
@@ -607,7 +606,8 @@ def post(self):
                 )
 
         with self.Session() as session:
-            try:
+            # try:
+            if True:
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/gcn.py:610]: will put this back.

File: skyportal/handlers/api/gcn.py [hunk 6]
```diff
@@ -664,14 +664,28 @@ async def get(self, dateobs=None):
...
+                type: string
+              description: |
+                Comma-separated string of `LocalizationTag`s to match against.
+            - in: query
+              name: localizationTagRemove
+              nullable: true
+              schema:
+                type: string
+              description: |
+                Comma-separated string of `LocalizationTag`s to filter out
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/gcn.py]: yes.

File: skyportal/models/user_notification.py (hunk 4 above)
Comment (mcoughlin) [reply skyportal/models/user_notification.py]: good point.

File: skyportal/models/user_notification.py [hunk 7]
```diff
@@ -710,23 +709,17 @@ def receive_after_flush(session, context):
...
                             if count_notices > 1:
-                                session.add(
-                                    UserNotification(
-                                        user=user,
-                                        text=f"New Notice for GCN Event *{target.dateobs}*, with Notice Type *{gcn.NoticeType(notice.notice_type).name}*",
-                                        notification_type="gcn_events",
-                                        url=f"/gcn_events/{str(target.dateobs).replace(' ','T')}",
-                                    )
-                                )
+                                text = f"New Notice for GCN Event *{target.dateobs}*, with Notice Type *{gcn.NoticeType(notice.notice_type).name}*"
```
Comment (guynir42) [inline skyportal/models/user_notification.py:712]:
```suggestion
                                text = f"New Notice for GCN Event *{target.dateobs}*, "
                                          f"with Notice Type *{gcn.NoticeType(notice.notice_type).name}*"
```

File: skyportal/models/user_notification.py [hunk 8]
```diff
@@ -710,23 +709,17 @@ def receive_after_flush(session, context):
...
                             else:
-                                session.add(
-                                    UserNotification(
-                                        user=user,
-                                        text=f"New GCN Event *{target.dateobs}*, with Notice Type *{gcn.NoticeType(notice.notice_type).name}*",
-                                        notification_type="gcn_events",
-                                        url=f"/gcn_events/{str(target.dateobs).replace(' ','T')}",
-                                    )
+                                text = f"New GCN Event *{target.dateobs}*, with Notice Type *{gcn.NoticeType(notice.notice_type).name}*"
```
Comment (guynir42) [inline skyportal/models/user_notification.py:714]:
```suggestion
                                text = f"New GCN Event *{target.dateobs}*, "
                                          f"with Notice Type *{gcn.NoticeType(notice.notice_type).name}*"
```

**Review by guynir42 — APPROVED** [review]
Looks much better, thanks! I suggest just breaking a couple of lines of text below:

**Pull request merged**